src/scanelf.d

```diff
@@ -42,152 +42,178 @@ void scanElfObjModule(void delegate(const(char)[] name, int pickAny) pAddSymbol,
     {
         printf("scanElfObjModule(%s)\n", module_name);
     }
-    const buf = base.ptr;
-    const buflen = base.length;
-    int reason = 0;
-    if (buflen < Elf32_Ehdr.sizeof)
+
+    void corrupt(int reason)
     {
-        reason = __LINE__;
-    Lcorrupt:
         error(loc, "corrupt ELF object module %s %d", module_name, reason);
-        return;
     }
-    if (memcmp(buf, elf, 4))
-    {
-        reason = __LINE__;
-        goto Lcorrupt;
-    }
-    if (buf[EI_VERSION] != EV_CURRENT)
+
+    if (base.length < Elf32_Ehdr.sizeof)
+        return corrupt(__LINE__);
+
+    static immutable ubyte[4] elf = [0x7F, 'E', 'L', 'F']; // ELF file signature
+    if (memcmp(base.ptr, elf.ptr, elf.length))
+        return corrupt(__LINE__);
+
+    if (base[EI_VERSION] != EV_CURRENT)
     {
-        error(loc, "ELF object module %s has EI_VERSION = %d, should be %d", module_name, buf[EI_VERSION], EV_CURRENT);
+        error(loc, "ELF object module %s has EI_VERSION = %d, should be %d", module_name, base[EI_VERSION], EV_CURRENT);
         return;
     }
-    if (buf[EI_DATA] != ELFDATA2LSB)
+    if (base[EI_DATA] != ELFDATA2LSB)
     {
         error(loc, "ELF object module %s is byte swapped and unsupported", module_name);
         return;
     }
-    if (buf[EI_CLASS] == ELFCLASS32)
+    if (base[EI_CLASS] == ELFCLASS32)
     {
-        Elf32_Ehdr* eh = cast(Elf32_Ehdr*)buf;
+        const eh = cast(const(Elf32_Ehdr)*) base.ptr;
         if (eh.e_type != ET_REL)
         {
             error(loc, "ELF object module %s is not relocatable", module_name);
             return; // not relocatable object module
         }
         if (eh.e_version != EV_CURRENT)
-            goto Lcorrupt;
+            return corrupt(__LINE__);
+
+        bool overflow;
+        const end = addu(eh.e_shoff, mulu(eh.e_shentsize, eh.e_shnum, overflow), overflow);
+        if (overflow || end > base.length)
+            return corrupt(__LINE__);
+
         /* For each Section
          */
-        for (uint u = 0; u < eh.e_shnum; u++)
+        const sections = (cast(const(Elf32_Shdr)*)(base.ptr + eh.e_shoff))[0 .. eh.e_shnum];
+        foreach (ref const section; sections)
         {
-            Elf32_Shdr* section = cast(Elf32_Shdr*)(buf + eh.e_shoff + eh.e_shentsize * u);
             if (section.sh_type == SHT_SYMTAB)
```

Comment:
What we have here, ladies and gentlemen, is an `if` statement that is 47 lines long. it is the only thing the enclosing `for` loop does, meaning there is no `else` and no code following it. Of course, that can only be known after desperately scrolling up and down in search for its ending brace, which is one of 5 (five) braces extending between lines 133 and 137. (Among other things there's even a function defined within - yes, four (4) levels of indentation inside. Not that there's anything wrong with it!)

But wait, there's more. One of those five (5) closing braces on lines 133-137 is _another_ `if` statement that is the only thing a `for` loop does.  In fact, the very condition within that `if` is a double negation which can be easier expressed in negated form:

````d
foreach (ref const sym; symbols)
{
    if (((sym.st_info >> 4) == STB_GLOBAL || (sym.st_info >> 4) == STB_WEAK) &&
                        sym.st_shndx != SHN_UNDEF) // not extern
    {
        ... code ...
    }
}
```

----->

````d
foreach (ref const sym; symbols)
{
    if ((sym.st_info >> 4) != STB_GLOBAL && (sym.st_info >> 4) != STB_WEAK ||
                        sym.st_shndx == SHN_UNDEF) // extern, nothing to do
        continue;
    ... code ...
}
```